File: frontend/lib/features/public_messages/data/models/public_message.dart
import 'package:equatable/equatable.dart';
import 'package:reallystick/features/public_messages/domain/entities/public_message.dart';

class PublicMessageDataModel extends Equatable {
  final String id;
  final String? habitId;
  final String? challengeId;
  final String creator;
  final String threadId;
  final String? repliesTo;
  final DateTime createdAt;
  final DateTime? updateAt;
  final String content;
  final int likeCount;
  final int replyCount;
  final bool deletedByCreator;
  final bool deletedByAdmin;
  final String? languageCode;

  const PublicMessageDataModel({
    required this.id,
    required this.habitId,
    required this.challengeId,
    required this.creator,
    required this.threadId,
    required this.repliesTo,
    required this.createdAt,
    required this.updateAt,
    required this.content,
    required this.likeCount,
    required this.replyCount,
    required this.deletedByCreator,
    required this.deletedByAdmin,
    required this.languageCode,
  });

  factory PublicMessageDataModel.fromJson(Map<String, dynamic> json) {
    return PublicMessageDataModel(
      id: json['id'] as String,
      habitId: json['habit_id'] as String?,
      challengeId: json['challenge_id'] as String?,
      creator: json['creator'] as String,
      threadId: json['thread_id'] as String,
      repliesTo: json['replies_to'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updateAt: json['updated_at'] is String
          ? DateTime.parse(json['updated_at'] as String)
          : null,
      content: json['content'] as String,
      likeCount: json['like_count'] as int,
      replyCount: json['reply_count'] as int,
      deletedByCreator: json['deleted_by_creator'] as bool,
      deletedByAdmin: json['deleted_by_admin'] as bool,
      languageCode: json['language_code'] as String?,
    );
  }

  PublicMessage toDomain() => PublicMessage(
        id: id,
        habitId: habitId,
        challengeId: challengeId,
        creator: creator,
        threadId: threadId,
        repliesTo: repliesTo,
        createdAt: createdAt,
        updateAt: updateAt,
        content: content,
        likeCount: likeCount,
        replyCount: replyCount,
        deletedByCreator: deletedByCreator,
        deletedByAdmin: deletedByAdmin,
        languageCode: languageCode,
      );

  @override
  List<Object?> get props => [
        id,
        habitId,
        challengeId,
        creator,
        threadId,
        repliesTo,
        createdAt,
        updateAt,
        content,
        likeCount,
        replyCount,
        deletedByCreator,
        deletedByAdmin,
        languageCode,
      ];
}

File: backend/src/features/public_discussions/helpers/public_message.rs
use sqlx::{postgres::PgQueryResult, PgConnection};
use uuid::Uuid;

use crate::features::public_discussions::structs::models::public_message::PublicMessage;

pub async fn get_reported_messages(
    conn: &mut PgConnection,
) -> Result<Vec<PublicMessage>, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
           SELECT pm.*
            FROM public_messages pm
            JOIN public_message_reports pmr ON pm.id = pmr.message_id;
        "#,
    )
    .fetch_all(conn)
    .await
}

pub async fn get_user_reported_messages(
    conn: &mut PgConnection,
    user_id: Uuid,
) -> Result<Vec<PublicMessage>, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
           SELECT pm.*
            FROM public_messages pm
            JOIN public_message_reports pmr ON pm.id = pmr.message_id
            WHERE pmr.reporter = $1;
        "#,
        user_id,
    )
    .fetch_all(conn)
    .await
}

pub async fn get_user_liked_messages(
    conn: &mut PgConnection,
    user_id: Uuid,
) -> Result<Vec<PublicMessage>, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
        SELECT pm.*
        FROM public_messages pm
        JOIN public_message_likes pml ON pm.id = pml.message_id
        WHERE pml.user_id = $1;
        "#,
        user_id,
    )
    .fetch_all(conn)
    .await
}

pub async fn get_user_written_messages(
    conn: &mut PgConnection,
    user_id: Uuid,
) -> Result<Vec<PublicMessage>, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
        SELECT *
        from public_messages
        WHERE creator = $1
        "#,
        user_id,
    )
    .fetch_all(conn)
    .await
}

pub async fn get_replies(
    conn: &mut PgConnection,
    message_id: Uuid,
) -> Result<Vec<PublicMessage>, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
        SELECT *
        from public_messages
        WHERE replies_to = $1
        "#,
        message_id,
    )
    .fetch_all(conn)
    .await
}

pub async fn get_first_public_messages_of_challenge(
    conn: &mut PgConnection,
    challenge_id: Uuid,
) -> Result<Vec<PublicMessage>, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
        SELECT *
        from public_messages
        WHERE challenge_id = $1 and replies_to is null and deleted_by_admin = false and deleted_by_creator = false
        "#,
        challenge_id,
    )
    .fetch_all(conn)
    .await
}

pub async fn get_first_public_messages_of_habit(
    conn: &mut PgConnection,
    habit_id: Uuid,
) -> Result<Vec<PublicMessage>, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
        SELECT *
        from public_messages
        WHERE habit_id = $1 and replies_to is null and deleted_by_admin = false and deleted_by_creator = false
        "#,
        habit_id,
    )
    .fetch_all(conn)
    .await
}

pub async fn get_public_message_by_id(
    conn: &mut PgConnection,
    id: Uuid,
) -> Result<Option<PublicMessage>, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
        SELECT *
        from public_messages
        WHERE id = $1
        "#,
        id,
    )
    .fetch_optional(conn)
    .await
}

pub async fn delete_public_message(
    conn: &mut PgConnection,
    public_message: &PublicMessage,
) -> Result<PgQueryResult, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
        UPDATE public_messages
        SET deleted_by_admin = $1, deleted_by_creator = $2
        WHERE id = $3
        "#,
        public_message.deleted_by_admin,
        public_message.deleted_by_creator,
        public_message.id,
    )
    .execute(conn)
    .await
}

pub async fn create_public_message(
    conn: &mut PgConnection,
    public_message: &PublicMessage,
) -> Result<PgQueryResult, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
        INSERT INTO public_messages (
            id,
            habit_id,
            challenge_id,
            creator,
            thread_id,
            replies_to,
            created_at,
            updated_at,
            content,
            like_count,
            deleted_by_creator,
            deleted_by_admin,
            language_code
        )
        VALUES ( $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13)
        "#,
        public_message.id,
        public_message.habit_id,
        public_message.challenge_id,
        public_message.creator,
        public_message.thread_id,
        public_message.replies_to,
        public_message.created_at,
        public_message.updated_at,
        public_message.content,
        public_message.like_count,
        public_message.deleted_by_creator,
        public_message.deleted_by_admin,
        public_message.language_code
    )
    .execute(conn)
    .await
}

pub async fn update_public_message(
    conn: &mut PgConnection,
    public_message: &PublicMessage,
) -> Result<PgQueryResult, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
        UPDATE public_messages
        SET updated_at = $1, content = $2
        WHERE id = $3
        "#,
        public_message.updated_at,
        public_message.content,
        public_message.id,
    )
    .execute(conn)
    .await
}

pub async fn update_public_message_like_count(
    conn: &mut PgConnection,
    public_message: &PublicMessage,
) -> Result<PgQueryResult, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
        UPDATE public_messages
        SET like_count = $1
        WHERE id = $2
        "#,
        public_message.like_count,
        public_message.id,
    )
    .execute(conn)
    .await
}

pub async fn update_public_message_reply_count(
    conn: &mut PgConnection,
    public_message: &PublicMessage,
) -> Result<PgQueryResult, sqlx::Error> {
    sqlx::query_as!(
        PublicMessage,
        r#"
        UPDATE public_messages
        SET reply_count = $1
        WHERE id = $2
        "#,
        public_message.reply_count,
        public_message.id,
    )
    .execute(conn)
    .await
}
